pyop2/array.py

```diff
@@ -0,0 +1,171 @@
+import enum
+
+import numpy as np
+import pyopencl
+from petsc4py import PETSc
+
+from pyop2.configuration import configuration
+from pyop2.offload_utils import _offloading as offloading
+from pyop2.types.access import READ, RW, WRITE, INC
+from pyop2.utils import verify_reshape
+
+
+class Status(enum.Enum):
+    """where is the up-to-date version?"""
+    ON_HOST = enum.auto()
+    ON_DEVICE = enum.auto()
+    ON_BOTH = enum.auto()
+
+
+ON_HOST = Status.ON_HOST
+ON_DEVICE = Status.ON_DEVICE
+ON_BOTH = Status.ON_BOTH
+
+
+class MirroredArray:
+    """An array that is available on both host and device, copying values where necessary."""
+
+    def __init__(self, data, dtype, shape):
+        # TODO This could be allocated lazily
+        self.dtype = np.dtype(dtype)
+        if data is None:
+            self._host_data = np.zeros(shape, dtype=dtype)
+        else:
+            self._host_data = verify_reshape(data, dtype, shape)
+        self.availability = ON_BOTH
+
+    @classmethod
+    def new(cls, *args, **kwargs):
+        if configuration["backend"] == "CPU_ONLY":
+            return CPUOnlyArray(*args, **kwargs)
+        elif configuration["backend"] == "OPENCL":
+            return OpenCLArray(*args, **kwargs)
+        else:
+            raise ValueError
+
+    @property
+    def kernel_arg(self):
+        # lazy but for now assume that the data is always modified if we access
+        # the pointer
+        return self.device_ptr if offloading else self.host_ptr
+
+        # if offloading:
+        #     # N.B. not considering MAX, MIN here
+        #     if access in {RW, WRITE, INC}:
+        #         return self.device_ptr
+        #     else:
+        #         return self.device_ptr_ro
+        # else:
+        #     # N.B. not considering MAX, MIN here
+        #     if access in {RW, WRITE, INC}:
+        #         return self.host_ptr
+        #     else:
+        #         return self.host_ptr_ro
+
+    @property
+    def is_available_on_device(self):
+        return self.availability in {ON_DEVICE, ON_BOTH}
+
+    def ensure_availability_on_device(self):
+        if not self.is_available_on_host:
+            self.host_to_device_copy()
+
+    @property
+    def is_available_on_host(self):
+        return self.availability in {ON_HOST, ON_BOTH}
+
+    def ensure_availability_on_host(self):
+        if not self.is_available_on_host:
+            self.device_to_host_copy()
+
+    @property
+    def vec(self):
```

Comment:
Subclasses would need to implement the right thing here. I have not tried to make my code work for `can_be_represented_as_petscvec` but I don't think it would require a radical rethink.